File: 3.breath-first-search/근우/5.binary-tree-level-order-traversal.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


from collections import deque
from typing import List, Optional


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        root.depth = 0
        que = deque([root])
        nodes = [root]
        while que:
            c = que.popleft()
            if c.left:
                c.left.depth = c.depth + 1
                nodes.append(c.left)
                que.append(c.left)
            if c.right:
                c.right.depth = c.depth + 1
                nodes.append(c.right)
                que.append(c.right)
        max_depth = nodes[-1].depth
        res = [[] for _ in range(max_depth + 1)]
        for n in nodes:
            res[n.depth].append(n.val)
        return res

File: 3.breath-first-search/동규/5.binary-tree-level-order-traversal.js
function levelOrder(root) {
  const res = [];
  if (root === null) return res;

  const queue = [];
  queue.push(root);
  while (queue.length !== 0) {
    let size = queue.length;
    let level = [];
    while (size--) {
      let cur = queue.shift();
      level.push(cur.val);
      if (cur.left) queue.push(cur.left);
      if (cur.right) queue.push(cur.right);
    }
    res.push(level.concat());
  }
  return res;
}

levelOrder([3, 9, 20, null, null, 15, 7]);
